### PR TITLE
[MkFit] Material derived in GeometryESProducer 

### PR DESCRIPTION
**Pull request by leonardogiannini**
#### PR description:

The material description for MkFit tracking was previously stored  as a look up table derived from the tracking ntuple in [MaterialEffects](https://github.com/cms-sw/cmssw/blob/master/RecoTracker/MkFitCore/src/MaterialEffects.cc). After this pull request the material is derived together with the Geometry in [MkFitGeometryESProducer](https://github.com/cms-sw/cmssw/blob/master/RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc)

The material is derived starting from the single modules (double sided modules excluded) in a r-z grid with bins of 1cm x 1cm. If more than one detector is present in a bin, a weighted average is used. The material is then expanded to the mkFit layers' boxes by filling the gaps with the material from the closest neighbor detector. 

In the end we obtain a grid in r-z with bins of 1cm x 1cm, which is used in the same way as the grid derived from the look up table.

The [MaterialEffects](https://github.com/cms-sw/cmssw/blob/master/RecoTracker/MkFitCore/src/MaterialEffects.cc) files are not removed, as they are still necessary in mkFit standalone, but will be removed later.

This implementation will also allow a smooth transition to the Phase 2 tracker.

#### PR validation:

- old r-z material grid (radiation length)
image

- new r-z material grid (radiation length)
image

-- NB: the material is the same where modules are present, and very similar in the layers boxes. The divisions were previously done by eye, while now we choose the closest neighbor automatically.

- MTV [validation](http://uaf-10.t2.ucsd.edu/~legianni/single-iter-checkmat/plots_compare_132x_1000ev/) shows compatible tracking results

### CODE REVIEW

**Comment by perrotta** [comment]
please test

File: RecoTracker/MkFitCore/interface/PropagationConfig.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  class TrackerInfo;
+
+  enum PropagationFlagsEnum {
+    PF_none = 0,
+    PF_use_param_b_field = 0x1,
+    PF_apply_material = 0x2,
+    PF_copy_input_state_on_fail = 0x4
+  };
+
+  struct PropagationFlags {
```
Comment (mmusich) [inline RecoTracker/MkFitCore/interface/PropagationConfig.h:15]:
```
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_2_X_2023-05-17-2300/src/RecoTracker/MkFitCore/interface/PropagationConfig.h:15:3: warning: 'PropagationFlags' defined as a struct here but previously declared as a class; this is valid, but may result in linker errors under the Microsoft C++ ABI [-Wmismatched-tags]
```

**Comment by leonardogiannini** [comment]
sorry for the delay. it should be good now.
@slava77

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc [hunk 2]
```diff
@@ -346,6 +399,82 @@ void MkFitGeometryESProducer::addTECGeometry(mkfit::TrackerInfo &trk_info) {
   edm::LogVerbatim("MkFitGeometryESProducer") << ostr.str();
 }
 
+void MkFitGeometryESProducer::findRZBox(const GlobalPoint &gp, float &rmin, float &rmax, float &zmin, float &zmax) {
+  float r = gp.perp(), z = gp.z();
```
Comment (mandrenguyen) [inline RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc:403]:
```suggestion
  float absr = std::fabs(gp.perp()), absz = std::fabs(gp.z());
```

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc [hunk 3]
```diff
@@ -346,6 +399,82 @@ void MkFitGeometryESProducer::addTECGeometry(mkfit::TrackerInfo &trk_info) {
...
+  if (std::fabs(r) > rmax)
+    rmax = std::fabs(r);
+  if (std::fabs(r) < rmin)
```
Comment (mandrenguyen) [inline RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc:406]: Could be `else if` if `rmax` is always bigger than `rmin`

**Review by mandrenguyen — CHANGES_REQUESTED** [review]

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc (hunk 2 above)
Comment (slava77) [reply RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc:403]: `r` is already non-negative

Comment (mandrenguyen) [reply RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc:403]: Ok, then please remove the `std::fabs` operating on `r`

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc (hunk 3 above)
Comment (slava77) [reply RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc:406]: > use `std::clamp`

`std::min` and `std::max` seem more appropriate clamp would look odd `rmax = clamp(r, rmax, r)`

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc (hunk 2 above)
Comment (leonardogiannini) [reply RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc]: review addressed

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc (hunk 3 above)
Comment (leonardogiannini) [reply RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc]: review addressed

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by mandrenguyen** [comment]
+reconstruction

**Comment by perrotta** [comment]
+1

**Pull request merged**